### DISCUSSION TITLE
Proposal: SustainabilityManager API

### DISCUSSION THREAD

**Discussion by lukeblevins** [microsoft/WindowsAppSDK#2525] [Ideas] [answered]
<img width="468" alt="image" src="https://user-images.githubusercontent.com/20365014/168884592-e8ed7d74-8c79-4848-8f1b-663802c88d59.png">

_There couldn't be a better time to add a new `Microsoft.Windows.System.Power.SustainabilityManager` API, providing responsible third-party developers with the same carbon-aware capabilities introduced with Windows Update in recent Windows 11 builds_

`SustainabilityManager` should allow developers to subscribe to events that fire when it is a more/less optimal time to perform computationally intensive operations. The event handler argument could include an estimated time when energy grid conditions will change again. 

This new API should follow the design pattern of others and contain a static `IsSupported()` method to check whether:
- The capability exists on the OS version
- It's not disabled by the user in Windows Settings 
- The PC is plugged-in
- There is network connectivity-to receive regional carbon intensity data from [electricityMap](https://app.electricitymap.org/map) or [WattTime](https://www.watttime.org/)
 
This class could also expose the long-awaited WinRT API for apps to enable EcoQoS (efficiency mode) on their processes. 

There are numerous positive implications for Microsoft, end users, and the planet if this technology is exposed to third party developers. Bringing this technology to the Windows App SDK could also yield a virtuous cycle of adoption by larger, widely renowned companies that are environmentally conscious.

**Comment by veikkoeeva** [comment]
Cross-referencing https://github.com/dotnet/runtime/issues/66417.

Probably good band-aid could be just having a vetted piece of .NET for setting and unsetting this flag. Non-.NET example at https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-setprocessinformation#examples. Maybe useful is a code example at https://github.com/dahall/Vanara/blob/f28c7c3d8b21d58a603b3077bcd432d8b811bb45/UnitTests/PInvoke/Kernel32/ProcessThreadsTests.cs#L477.

**Comment by lukeblevins** [comment, marked as answer]
### An update for this proposal - almost 2 years later

In mid-2022, I opened this proposal after hearing news that in Windows 11, regular updates would become [carbon-aware](https://support.microsoft.com/windows/windows-update-is-now-carbon-aware-a53f39bc-5531-4bb1-9e78-db38d7a6df20) with 22H2. In case you missed it, this improvement makes Windows Update take into account the forecasted carbon-intensity of sources on a user's regional power grid when it determines the optimal time to schedule installation attempts. I anticipated that the new data sources which became leveraged by the OS could be more broadly valuable if exposed to third-party app developers through a new API delivered through the Windows App SDK. 

This discussion was eventually closed by me last year after WinAppSDK development appeared to stall and it was not a pressing need for most developers compared to other feature gaps. 

However, I am happy to see an API for this feature has very recently been announced in the [notes](https://blogs.windows.com/windows-insider/2024/02/08/announcing-windows-11-insider-preview-build-26052-canary-and-dev-channels/) of a recent Insider Preview build (26000-series) as Power Grid Forecast API:
> Starting with this build, we are introducing the Power Grid Forecast API. This API empowers app developers to optimize app behavior, minimizing environmental impact by shifting background tasks to times when more renewable energy is available in the local electrical grid. The result is lower carbon emissions for app users. Please note that forecast quality varies by region and is not available globally. While the API is available in this build, we are still working on public documentation, sample code, and we have a couple of bug fixes checked in. All of this should make its way to Insiders in the coming weeks.

Documentation for this pre-release API can so far be found as [`PowerGridForecast`](https://learn.microsoft.com/uwp/api/windows.devices.power.powergridforecast?view=winrt-insider) and [`PowerGridData`](https://learn.microsoft.com/uwp/api/windows.devices.power.powergriddata?view=winrt-insider).

Please note that this feature is not exposed to developers using Windows App SDK, which means it is currently coupled to the platform itself, and only available on build 26058 (Preview) or newer. The API is also subject to change, and I lack any knowledge or guarantees beyond this exciting possibility I noticed as a user.

Anyhow, I'd say this ask is now completed!
